Extend App Protect docs and example (#1039)
nginxinc/kubernetes-ingress: docs-web/app-protect/troubleshooting.md

# Troubleshooting

This document describes how to troubleshoot problems with the Ingress Controller with the [App Protect](/nginx-app-protect/) module enabled.

For general troubleshooting of the Ingress Controller, check the general [troubleshooting](/nginx-ingress-controller/troubleshooting/) documentation.

For additional troubleshooting of the App Protect module itself, check the [troubleshooting](/nginx-app-protect/troubleshooting/) guide in the App Protect module documentation. 

## Potential Problems

> The table below categorizes some potential problems with the App Protect integration with the Ingress Controller you may encounter and suggests how to troubleshoot those problems using one or more methods from the next section.

Reviewer: ```suggestion
The table below categorizes some potential problems with the Ingress Controller when App Protect module is enabled. It suggests how to troubleshoot those problems, using one or more methods from the next section.
```

```eval_rst
.. list-table::
   :header-rows: 1

   * - Problem area
     - Symptom
     - Troubleshooting method
     - Common cause
   * - Start.
     - The Ingress Controller fails to start.
     - Check the logs.
     - Misconfigured APLogConf or APPolicy.
   * - APLogConf, APPolicy or Ingress Resource.
     - The configuration is not applied.
     - Check the events of the APLogConf, APPolicy and Ingress Resource, check the logs.
     - APLogConf or APPolicy is invalid.
   * - NGINX.
     - The Ingress Controller NGINX verification timeouts while starting for the first time or while reloading after a change.
     - Check the logs for ``Unable to fetch version: X`` message.
     - Too many Ingress Resources with App Protect enabled. Check the `NGINX fails to start/reload section <#nginx-fails-to-start-or-reload>`_ of the Known Issues. 
```

## Troubleshooting Methods

> ### Checking the Ingress Controller and App Protect logs

Reviewer 2: ```suggestion
### Check the Ingress Controller and App Protect logs
```

> App Protect logs are part of the Ingress Controller logs when the module is enabled. To check the Ingress Controller logs, follow the steps of [Checking the Ingress Controller Logs](/nginx-ingress-controller/troubleshooting/#checking-the-ingress-controller-logs) of the Troubleshooting guide.

Author: I am not sure, those logs come from the module, let me check to see if I can add something

```
Note that in the events section, we have a `Normal` event with the `AddedOrUpdated` reason, which informs us that the configuration was successfully applied.

### Checking events of APPolicy

After you create or update an APPolicy, you can immediately check if the NGINX configuration was successfully applied by NGINX:
```
$ kubectl describe appolicy dataguard-alarm
Name:         dataguard-alarm
Namespace:    default
. . . 
Events:
  Type    Reason          Age    From                      Message
  ----    ------          ----   ----                      -------
  Normal  AddedOrUpdated  2m25s  nginx-ingress-controller  AppProtectPolicy default/dataguard-alarm was added or updated
```
Note that in the events section, we have a `Normal` event with the `AddedOrUpdated` reason, which informs us that the configuration was successfully applied.
 
## Running App Protect in Debug Mode

Running the Ingress Controller in debug mode will also run the App Protect module in debug mode. Follow the steps of [Running NGINX in the Debug Mode](/nginx-ingress-controller/troubleshooting/#running-nginx-in-the-debug-mode).

## Known Issues

> When using the Ingress Controller with the App Protect module, the following issues might happen depending on the number of Ingress Resources with App Protect enabled in your cluster:

Reviewer 3: @jputrino those issues are possible side-effects of the integration. I'm not sure that saying `have been reported` is applicable here - we know about them.

Reading the issue list it seems we have two issues:
* *NGINX configuration skew* (replicas having different configuration. Right now this issue is part of the `NGINX reload times increase` section. I don't think `NGINX reload times increase` is an issue, rather it is a side effect that causes other issues ) 
* *NGINX fails to start or reload*

Perhaps it makes sense to organize the known issues list like below (the existing text can be reused, just a slightly different structure):

## Known Issues

A few sentences about NGINX reload times increase, which is the side-effect of the integration.

### NGINX configuration Skew
... 

### NGINX fails to start or reload
...

> When using the Ingress Controller with the App Protect module, the following issues might happen depending on the number of Ingress Resources with App Protect enabled in your cluster:

Author: Let me know what to do here @pleshakov @jputrino

> When using the Ingress Controller with the App Protect module, the following issues might happen depending on the number of Ingress Resources with App Protect enabled in your cluster:

Reviewer 2: @Rulox I'd follow @pleshakov's recommendation.

### NGINX reload times increase

> The Ingress Controller reloads NGINX whenever there is a change that requires NGINX to reload to apply a new configuration. Without the App Protect module enabled, usual reload times are around 150ms. If App Protect module is enabled and is being used by any number of Ingress Resources, these reloads might take a few seconds instead. If you are running more than one instance of the Ingress Controller, because there is no order on how the Ingress Controller processes the Kubernete Reources, it might occur that different instances have different NGINX configurations until all of them have finished with the reloads.

Reviewer 2: ```suggestion
When you make a change that requires NGINX to apply a new configuration, the Ingress Controller reloads NGINX automatically. Without the App Protect module enabled, usual reload times are around 150ms. If App Protect module is enabled and is being used by any number of Ingress Resources, these reloads might take a few seconds instead. 

> **Note**: If you are running more than one instance of the Ingress Controller, the extended reload time may cause the NGINX configuration of your instances to be out of sync. This can occur because there is no order imposed on how the Ingress Controller processes the Kubernetes Resources. The configurations will be the same after all instances have completed the reload.
```

> In order to reduce these inconsistencies, we recommend not to apply changes to resources handled by the Ingress Controller at the same time. 

Reviewer 2: ```suggestion
In order to reduce these inconsistencies, we advise that you do not apply changes to multiple resources handled by the Ingress Controller at the same time. 
```

### NGINX fails to start or reload

> The first time the Ingress Controller starts or whenever there is a change that requires reloading NGINX, the Ingress Controller will verify if the reload was successful. There is a timeout for this verification. When App Protect is enabled, this timeout is 20 seconds (instead of 4 seconds without the module). 

Reviewer 2: ```suggestion
The first time the Ingress Controller starts, or whenever there is a change that requires reloading NGINX, the Ingress Controller will verify if the reload was successful. The timeout for this verification is normally 4 seconds. When App Protect is enabled, this timeout is 20 seconds. 
```

> This timeout should be more than enough to verify configurations. However, when numerous Ingress Resources with App Protect enabled are handled by the Ingress Controller at the same time (for instance, if you apply a large amount of Ingress Resources at once, or when the Ingress Controller runs for the first time a cluster where the Ingress Resources with App Protect enabled are already present), this timeout might not be enough.

Reviewer: ```suggestion
This timeout should be more than enough to verify configurations. However, when numerous Ingress Resources with App Protect enabled are handled by the Ingress Controller at the same time (for instance, if you apply a large amount of Ingress Resources at once, or when the Ingress Controller runs for the first time in a cluster where the Ingress Resources with App Protect enabled are already present), this timeout might not be enough.
```

> This timeout should be more than enough to verify configurations. However, when numerous Ingress Resources with App Protect enabled are handled by the Ingress Controller at the same time (for instance, if you apply a large amount of Ingress Resources at once, or when the Ingress Controller runs for the first time a cluster where the Ingress Resources with App Protect enabled are already present), this timeout might not be enough.

Reviewer 2: ```suggestion
This timeout should be more than enough to verify configurations. However, when numerous Ingress Resources with App Protect enabled are handled by the Ingress Controller at the same time, you may find that you need to extend the timeout further.  Examples of when this might be necessary include:

- You need to apply a large amount of Ingress Resources at once.
- You are running the Ingress Controller for the first time in a cluster where the Ingress Resources with App Protect enabled are already present.
```
